pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,593 @@
+	return r.reconcile(req, r, log)
+}
+
+func (r *ForkliftPopulatorReconciler) reconcile(req reconcile.Request, populator populatorController, log logr.Logger) (reconcile.Result, error) {
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if pvc.Status.Phase == corev1.ClaimBound {
+		r.log.Info("PVC is bound, skipping...", "pvc", pvc.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// We first perform the common reconcile steps.
+	// We should only continue if we get a valid PVC'
+	pvcPrime, err := r.reconcileCommon(pvc, populator, log)
+	if err != nil || pvcPrime == nil {
+		return reconcile.Result{}, err
+	}
+
+	r.log.V(1).Info("reconciling PVC prime", "pvc", pvcPrime.Name)
+
+	// Each populator reconciles the target PVC in a different way
+	if cc.IsUnbound(pvc) || !cc.IsPVCComplete(pvc) {
+		return populator.reconcileTargetPVC(pvc, pvcPrime)
+	}
+
+	return r.reconcileCleanup(pvcPrime)
+}
+
+// TODO(benny) rename
+func (r *ForkliftPopulatorReconciler) reconcileCommon(pvc *corev1.PersistentVolumeClaim, populator populatorController, log logr.Logger) (*corev1.PersistentVolumeClaim, error) {
+	if pvc.DeletionTimestamp != nil {
+		log.V(1).Info("PVC being terminated, ignoring")
+		return nil, nil
+	}
+
+	log.V(1).Info("reconciling PVC prime")
+	pvcPrime, err := r.getPVCPrime(pvc)
+	if err != nil {
+		return nil, err
+	}
+	if pvcPrime != nil {
+		return pvcPrime, nil
+	}
+
+	dataSourceRef := pvc.Spec.DataSourceRef
+
+	if !isPVCForkliftKind(pvc) {
+		log.V(1).Info("reconciled unexpected PVC, ignoring")
+		return nil, nil
+	}
+
+	// TODO: Remove this check once we support cross-namespace dataSourceRef
+	if dataSourceRef.Namespace != nil {
+		log.V(1).Info("cross-namespace dataSourceRef not supported yet, ignoring")
+		return nil, nil
+	}
+
+	populationSource, err := populator.getPopulationSource(pvc)
+	if populationSource == nil {
+		return nil, err
+	}
+
+	ready, nodeName, err := claimReadyForPopulation(context.TODO(), r.client, pvc)
+	if !ready || err != nil {
+		return nil, err
+	}
+
+	if cc.IsUnbound(pvc) {
+		_, err := r.createPVCPrime(pvc, populationSource, nodeName != "", populator.updatePVCForPopulation)
+		if err != nil {
+			r.recorder.Eventf(pvc, corev1.EventTypeWarning, errCreatingPVCPrime, err.Error())
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+func (r *ForkliftPopulatorReconciler) getPopulationSource(pvc *corev1.PersistentVolumeClaim) (client.Object, error) {
+	switch pvc.Spec.DataSourceRef.Kind {
+	case "OvirtVolumePopulator":
+		return &v1beta1.OvirtVolumePopulator{}, nil
+	case "OpenstackVolumePopulator":
+		return &v1beta1.OpenstackVolumePopulator{}, nil
+	default:
+		return nil, fmt.Errorf("unknown populator type %T", pvc.Spec.DataSourceRef.Kind)
+	}
+}
+
+func isPVCForkliftKind(pvc *corev1.PersistentVolumeClaim) bool {
```

Comment:
So I don't understand this function, what I am expecting based on the name is that we check the `dataSourceRef` to have the right api group, and that it has a name for the populator, and that the Kind is in the supportedPopulators map. But when I look at this:

```go
if dataSourceRef.APIGroup != nil &&
  *dataSourceRef.APIGroup == apiGroup &&
  dataSourceRef.Name != "" {
    return false
}
```
It will pass if the APIGroup is nil or if the APIGroup doesn't match. And it fails if the name is a value. Is that the correct behavior we want? Seems strange to me.

Reply:
you're right and I couldn't understand how it worked until I saw that's because of apiGroup const